### ISSUE TITLE
cmake fail to set CXX_FLAGS correctly with independent external kokkos 

### ISSUE THREAD

**Issue by RichardYCJ** [dealii/dealii#14179]
It seems that the cmake of dealii does not handle the compile/link flags correctly from independent external kokkos. I encountered the following errors while building with spack:

```
==> dealii: Executing phase: 'build'
==> [2022-08-08-13:57:25.244577] 'ninja' '-j64' '-v'
[1/1314] /PATH/TO/mpicxx -SOME_FLAGS... -march=znver2 -mtune=znver2 -Wno-unused-local-typedefs -march=znver2;-mtune=znver2 ...
FAILED: cmake/scripts/CMakeFiles/expand_instantiations_exe.dir/expand_instantiations.cc.o 
/PATH/TO/mpicxx -SOME_FLAGS... -march=znver2 -mtune=znver2 -Wno-unused-local-typedefs -march=znver2;-mtune=znver2 ...
/usr/bin/ld: /usr/lib/gcc/x86_64-linux-gnu/9/../../../x86_64-linux-gnu/Scrt1.o: in function `_start':
(.text+0x24): undefined reference to `main'
collect2: error: ld returned 1 exit status
/bin/sh: 1: -mtune=znver2: not found
ninja: build stopped: subcommand failed.
```
Detailed logs can be seen [here](https://github.com/dealii/dealii/files/9279986/spack-build-out.txt).

I noticed there're **two replicated march/mtune flags** and one **separated by semi-colon**, which is not recognized as separators by gcc.

I've checked the compile/link flags of kokkos:
```
$<$<COMPILE_LANGUAGE:CXX>:-march=znver2;-mtune=znver2>
$<$<LINK_LANGUAGE:CXX>:-DKOKKOS_DEPENDENCE;-march=znver2;-mtune=znver2>
```

and according to what `FindKOKKOS.cmake` [does](https://github.com/dealii/dealii/blob/master/cmake/modules/FindKOKKOS.cmake), it only keeps the content of the generator expression by regex. **This would be a semi-colon separated string if there are multiple values.**

**The workaround is also simple, just leave the generator expression untouched.** I can start a pull request if you guys think this is a proper solution.

**Comment by RichardYCJ** [comment]
Besides, does `kokkos/arborx` need any extra configuration if built with cuda? I see from #12894 that part of `cuda::MatrixFree` is implemented using kokkos but the spack uses the default configuration without cuda aware.

**Comment by masterleinad** [comment]
At the moment, `ArborX` only uses the Host execution space and it's the only dependency using `Kokkos`. Thus, there is not much of a point configuring the `Kokkos` library used with `Cuda` support.
There are plans to let the CUDA implementation of the `MatrixFree` backend use `Kokkos`, also see #13549, but that pull request is very much work in progress.
As for the flags, there are some changes in https://github.com/dealii/dealii/pull/13549/files#diff-8f04862f3dc9a9d9f3a8acda350fbf12f1ba3ad2447f1c126b6512cdaf4a9e76 but if you think that you have a better solution, a pull request is very much welcome! Note that we want to be able to just use `target_link_libraries` but that needs some overhaul in our build system.

**Comment by RichardYCJ** [comment]
> As for the flags, there are some changes in https://github.com/dealii/dealii/pull/13549/files#diff-8f04862f3dc9a9d9f3a8acda350fbf12f1ba3ad2447f1c126b6512cdaf4a9e76

Seems right for me to address this problem. Though I think to directly use the compile/link flag offerred by kokkos is more straightforward, since this is safer **in case the language is not CXX** (almost impossible, like build in an project with extra language support?)

> Note that we want to be able to just use target_link_libraries but that needs some overhaul in our build system.

Totally agreed! Migrate to modern cmake would make the code more clear and expressive, but the risk of refactoring is also big.

Hope this patch could soon be availale in spack, feel free to ask if there's anything I could do (new to spack development :-)

**Comment by masterleinad** [comment]
Fixed by #14203.
